Right-click on links doesn't work in WebKit with Edit Menu enabled

mwyrzykowski opened the pull request:
#### 5f6b3147994b8b806eaee6cf67f54e4f92ef2b48
<pre>
Right-click on links doesn&apos;t work in WebKit with Edit Menu enabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=241395">https://bugs.webkit.org/show_bug.cgi?id=241395</a>

Reviewed by Wenson Hsieh.

UIKit is installing another UIContextMenuInteraction which was resulting in
two UIContextMenuInteraction instances installed on the view. Right-click on links
and some images showed an incorrect &apos;Select All&apos; menu only.

To avoid this, use the provided SPIs from UIKit which result in a single installed
UIContextMenuInteraction handler.

* Source/WebKit/Platform/spi/ios/UIKitSPI.h
Stage changes for non-internal builds.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
Add convenience property for when USE(UICONTEXTMENU) is not defined.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView textInteractionAssistantContextMenuInteraction]):
Helper for checking respondsToSelector before attempting to call SPI method.

(-[WKContentView setUpInteraction]):
The textSelectionAssistant needs to be created before registerPreview is called
in order to decided if we should use the SPI.

(-[WKContentView gestureRecognizer:shouldRequireFailureOfGestureRecognizer:]):
(-[WKContentView gestureRecognizer:shouldBeRequiredToFailByGestureRecognizer:]):
(-[WKContentView deferringGestureRecognizer:shouldDeferOtherGestureRecognizer:]):
(-[WKContentView imageAnalysisGestureDidTimeOut:]):
(-[WKContentView contextMenuInteraction]):
Use helper method which returns the current UIContextMenuInteraction handler.

(-[WKContentView _registerPreview]):
Use the UIKit provided UIContextMenuInteraction if it exists and set the delegate
to the same as what we would have passed to UIContextMenuInteraction&apos;s initializer.

Otherwise use the previous shipping behavior.

Canonical link: <a href="https://commits.webkit.org/251828@main">https://commits.webkit.org/251828@main</a>
</pre>

mwyrzykowski commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -1071,6 +1079,9 @@ - (void)setUpInteraction
     [self addGestureRecognizer:_touchActionGestureRecognizer.get()];
 
+    // FIXME: This should be called when we get notified that loading has completed.
+    [self setUpTextSelectionAssistant];
```
I couldn't observe any issue moving this call up 13 lines in this method, as far as I can tell it just initializes or resets the UIWKTextInteractionAssistant

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -183,6 +183,16 @@
+@interface WKContentView (ContextMenuAdditions)
+@property (nonatomic, readonly) UIContextMenuInteraction *contextMenuInteraction;
+@end
+#endif
```
This is being imported on non-internal builds.

Instead, it should go into the UIKitSPI.h inside WebKit when `USE(APPLE_INTERNAL_SDK)` is set, and fall back to forward declarations for non-internal builds.

mwyrzykowski replied:
Looks like that header is already imported in UIKitSPI.h, so I can remove it from here

whsieh commented on Source/WTF/wtf/PlatformUse.h:
```diff
@@ -279,6 +279,7 @@
 
 #if PLATFORM(IOS) || PLATFORM(MACCATALYST)
 #define USE_UICONTEXTMENU 1
+#define USE_UIKIT_CONTEXT_MENU_INTERACTION (defined(UIKIT_SUPPORTS_CONTEXT_MENU_INTERACTION) && UIKIT_SUPPORTS_CONTEXT_MENU_INTERACTION)
```
I think you can also just use the existing HAVE_TEXT_INTERACTION_WITH_CONTEXT_MENU_INTERACTION flag as well, instead of adding a new one?

mwyrzykowski replied:
oh that's perfect, I'll use that instead, thank you

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -183,6 +183,12 @@
+@property (nonatomic, readonly) UIContextMenuInteraction *contextMenuInteraction;
+@end
+#endif
+
```
Nit - I think you can just omit this.

Edit: I see, it's defined in `WKInteractionPreview`, rather than `ContextMenuAdditions`. In that case, I would rename the `ContextMenuAdditions` category here to `WKInteractionPreview`, to be consistent with that.

mwyrzykowski committed:
Right-click on links doesn't work in WebKit with Edit Menu enabled
https://bugs.webkit.org/show_bug.cgi?id=241395

Reviewed by Wenson Hsieh.

UIKit is installing another UIContextMenuInteraction which was resulting in
two UIContextMenuInteraction instances installed on the view. Right-click on links
and some images showed an incorrect 'Select All' menu only.

To avoid this, use the provided SPIs from UIKit which result in a single installed
UIContextMenuInteraction handler.

* Source/WebKit/Platform/spi/ios/UIKitSPI.h
Stage changes for non-internal builds.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
Add convenience property for when USE(UICONTEXTMENU) is not defined.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView textInteractionAssistantContextMenuInteraction]):
Helper for checking respondsToSelector before attempting to call SPI method.

(-[WKContentView setUpInteraction]):
The textSelectionAssistant needs to be created before registerPreview is called
in order to decided if we should use the SPI.

(-[WKContentView gestureRecognizer:shouldRequireFailureOfGestureRecognizer:]):
(-[WKContentView gestureRecognizer:shouldBeRequiredToFailByGestureRecognizer:]):
(-[WKContentView deferringGestureRecognizer:shouldDeferOtherGestureRecognizer:]):
(-[WKContentView imageAnalysisGestureDidTimeOut:]):
(-[WKContentView contextMenuInteraction]):
Use helper method which returns the current UIContextMenuInteraction handler.

(-[WKContentView _registerPreview]):
Use the UIKit provided UIContextMenuInteraction if it exists and set the delegate
to the same as what we would have passed to UIContextMenuInteraction's initializer.

Otherwise use the previous shipping behavior.

Canonical link: https://commits.webkit.org/251828@main
